USSF-ORBIT/ussf-portal-cms
# feat: Add tracking & logging to existing lists

# SC-366 , SC-365

## Proposed changes

This PR adds a [wrapper helper function](https://github.com/USSF-ORBIT/ussf-portal-cms/pull/38/files#diff-9bc66bf8fc3240de8ab5ef583bf8ff0bb86e2b80e6f0fd8c6d6505c63f2aa195) for Keystone lists that adds the following to list schema:
- `createdAt, createdBy, updatedAt, updatedBy` fields that are set automatically
- create Event records for all mutations performed on the list using [Keystone hooks](https://keystonejs.com/docs/apis/hooks#after-operation)

It also adds the above to all existing lists: User, Collection, and Bookmark.

Other:
- Minor change to the session object to populate `session.itemId, session.listKey` which is used for the by-tracking fields
- Split up `schema.test.ts` into individual test files for each list
- Started a testHelpers file that exports functions used for unit tests

---

## Reviewer notes

Test cases:
- Log in as a CMS user (`cmsuser / cmsuserpass`) and verify that:
    - you can make changes to your own user display name, and see the updatedAt/updatedBy values update.
    - you cannot view any Events on the Events list page
- Log in as a CMS admin (`cmsadmin / cmsadminpass`) and verify that:
    - you can make changes to any existing items and see the updatedAt/updatedBy values update.
    - you can create a new Bookmark or Collection, and see the correct values under createdAt/createdBy
    - after the above, you can view the corresponding Events on the Events list page

## Setup

To test locally, you need all services running:

```
yarn portal:up
```

and the CMS:
```
yarn install
yarn dev
```

---

## Code review steps

### As the original developer, I have

- For any new migrations/schema changes:

### As code reviewer(s), I have
